Feature/identity UI

evgeniikozhanov opened the pull request.

fedosov commented on app/components/wallet/Account.client.tsx:
```diff
@@ -162,6 +165,13 @@ export default function Account({ pair }: IProps) {
       {dialogElements}
       <AddressItem address={pair.address} />
+      <Icon
+        className="bp4-icon bp4-intent-success"
```
Константы

evgeniikozhanov replied:
fixed

fedosov commented on app/components/wallet/Account.client.tsx:
```diff
@@ -162,6 +165,13 @@ export default function Account({ pair }: IProps) {
       <AddressItem address={pair.address} />
+      <Icon
+        className="bp4-icon bp4-intent-success"
+        style={{'position': 'absolute', 'right': '0.3rem', 'top': '2.5rem', 'cursor': 'pointer'}}
```
Юзай tailwind

evgeniikozhanov replied:
ok, thanks

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,189 @@
+    <Dialog isOpen={isOpen} usePortal={true} onOpening={handleOnOpening} title="Identity" onClose={onClose} className="w-[90%] sm:w-[640px]">
+      <div className={`${Classes.DIALOG_BODY} flex flex-col gap-3`}>
+        {!data.isRegistrar && (
+            <Select2<IPalletIdentityRegistrarInfo>
```
Странно выглядит, тут нет ошибки?

evgeniikozhanov replied:
it is not necessary, I removed it

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,189 @@
+    }
+    const registrarInfo = (await api.query.identity.identityOf(registrarAccount.account)).toHuman();
+    registrarAccount = {...registrarAccount, ...(registrarInfo as IPalletIdentityRegistrarInfo)};
+    console.log('registrarAccount: ', registrarAccount);
```
Логирование

evgeniikozhanov replied:
wait, wait, it is only a draft)

evgeniikozhanov replied:
this block was removed

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,189 @@
+    }
+    setIsLoading(true);
+    try {
+    //   const tx = api.tx.miningPool.addMemberSelf(data.poolToJoin);
```
Мертвый код

evgeniikozhanov replied:
I started using it now)

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,189 @@
+
+  useEffect(() => {
+    setCanSubmit(api !== undefined);
+  }, [api, {}]);
```
Что значит {}?

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,189 @@
+    email: IRegistrarInfoItem;
+    web: IRegistrarInfoItem;
+    twitter: IRegistrarInfoItem;
+    // image: Any;
```
Мертвый код

evgeniikozhanov replied:
I started using it now)

evgeniikozhanov replied:
I have no example for it, therefore I have no idea how to deal with it

evgeniikozhanov replied:
my suggestion is that it is base64 data... or url...

fedosov commented on app/components/common/CandidateCards.tsx:
```diff
@@ -0,0 +1,95 @@
+type IProps = {
+  regIndex: number;
+  pair: KeyringPair;
+  dateMonthAgo: Date;
```
Не понятно зачем сюда передавать это значение, надо подписать

evgeniikozhanov replied:
I added a comment for this row

fedosov commented on app/components/common/CandidateCards.tsx:
```diff
@@ -0,0 +1,95 @@
+              <UserCard registrarInfo={candidateInfo} />
+              <Button
+                intent={Intent.PRIMARY}
+                disabled={false}
```
Это зачем?

evgeniikozhanov replied:
no ideas, removed

fedosov commented on app/components/common/UserCard.tsx:
```diff
@@ -0,0 +1,60 @@
+      <div className="block mb-2">
+        <div className="flex items-center gap-x-6">
+          {registrarInfo.info?.image?.Raw && <img className="h-16 w-16 rounded-full" src={registrarInfo.info.image.Raw as string} alt="" />}
+          {!registrarInfo.info?.image?.Raw && <Icon icon="person" iconSize={IconSize.LARGE * 3} title="no foto"></Icon>}
```
Сложно выглядит, можно тут использовать тернарный оператор? `registrarInfo.info?.image?.Raw ? ... : ...`

evgeniikozhanov replied:
yes, thank you for your suggestion! done

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,312 @@
+    isRegistrar: false,
+    registrarData: null,
+    candidateInfo: {} as ICandidateInfo,
+    dateMonthAgo: new Date(),
```
Тоже непонятно. Что это за дата «месяц назад» и почему она равна текущей дате?

evgeniikozhanov replied:
I changed it, hope it becomes more clear

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,312 @@
+                />
+                <Button
+                  intent={Intent.PRIMARY}
+                  disabled={isLoading || !canSubmit}
```
`canSubmit` всего в одном месте используется и сравнивает api с undefined, давай это упростим

evgeniikozhanov replied:
ok, I changed it, thanks

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,312 @@
+            )}
+          </>
+        )}
+        {!dataState.isRegistrar && Boolean(pair.meta.isInjected) && <div>You already have identity</div>}
```
Тут странное условие — если я не регистратор, то у меня уже есть Identity. Это не так, Identity мы должны еще запросить у Registrar, а он его должен одобрить, только тогда у нас появится Identity

evgeniikozhanov replied:
The condition is: if I'm not a registrar AND I have an identity. But I'm using a new variable for this purpose, hope it's clear now.

fedosov commented on app/components/wallet/Account.client.tsx:
```diff
@@ -233,6 +236,12 @@ export default function Account({ pair }: IProps) {
       {dialogElements}
       <AddressItem address={pair.address} />
+      <Icon
+        className={`${Classes.ICON} ${Classes.INTENT_SUCCESS} absolute cursor-pointer right-1 top-9`}
```
Давай сделаем хотя бы серой иконку, если нет identity, иначе она бесполезная — всегда показывает что у нас есть identity

evgeniikozhanov replied:
done

fedosov commented on app/utils/events.ts:
```diff
@@ -0,0 +1,42 @@
+  queryEvents: QueryResult<EventsData, EventsVars>
+): Promise<IPalletIdentityRegistrarInfo[]> {
+  const candidateInfoArray: IPalletIdentityRegistrarInfo[] = [];
+  if (queryEvents.data !== undefined && registrarIndex !== undefined && api) {
```
Давай в таких случаях инвертировать условие и делать return. Так уменьшится идентация у метода

evgeniikozhanov replied:
I like it! thanks!

fedosov commented on app/utils/events.ts:
```diff
@@ -0,0 +1,42 @@
+    let candidateAddress: string;
+    let candidateInfo: IPalletIdentityRegistrarInfo;
+    let addressAndRegIndex;
+    for (let i = 0; i < queryEvents.data.getEvents.objects.length; i++) {
```
`for (const item of queryEvents.data.getEvents.objects) {`

evgeniikozhanov replied:
thanks, I adjusted the code

fedosov reviewed (changes requested):
Сейчас все адреса помечаются как прошедшие верификацию. Так не должно быть.

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,307 @@
+        onClick={handleClick as MouseEventHandler}
+        onFocus={handleFocus}
+        roleStructure="listoption"
+        text={registrar.fee}
```
Здесь нужно отформатировать fee, можно использовать компонент `FormattedAmount`

fedosov reviewed (changes requested):
При запросе judgement получаю сообщение «Account is locked», хотя это аккаунт из расширения браузера (т.е. injected), такого не должно быть сообщения

evgeniikozhanov commented:
> При запросе judgement получаю сообщение «Account is locked», хотя это аккаунт из расширения браузера (т.е. injected), такого не должно быть сообщения

should be fixed, could you please check it again?

evgeniikozhanov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -0,0 +1,307 @@
+        active={modifiers.active}
+        disabled={modifiers.disabled}
+        key={registrar.account}
+        label={registrar.account}
```
but `api.query.identity.registrars()` returns only addresses and fees... Should I call the `getIdentity()` method for each of registrars?

fedosov replied:
Да, стоит попробовать запросить. По идее регистраторов всегда будет ограниченное количество. Если запросы послать параллельно, должно очень быстро отрабатывать

evgeniikozhanov replied:
done

fedosov commented:
Попробовал запросить judgement, получил вот такую ошибку:

```
createType(Call):: Call: failed decoding identity.setIdentity:: Struct: failed on args: {"info":"{\"additional\":\"Vec<(Data,Data)>\",\"display\":\"Data\",\"legal\":\"Data\",\"web\":\"Data\",\"riot\":\"Data\",\"email\":\"Data\",\"pgpFingerprint\":\"Option<[u8;20]>\",\"image\":\"Data\",\"twitter\":\"Data\"}"}:: Struct: failed on info: {"additional":"Vec<(Data,Data)>","display":"Data","legal":"Data","web":"Data","riot":"Data","email":"Data","pgpFingerprint":"Option<[u8;20]>","image":"Data","twitter":"Data"}:: Struct: failed on display: {"_enum":{"None":"Null","Raw":"Bytes","BlakeTwo256":"H256","Sha256":"H256","Keccak256":"H256","ShaThree256":"H256"}}:: Unable to decode Data, invalid indicator byte 116
```

Выбирал второго регистратора (Caldera), заполнял поля display name "test identity" и поле discord "fedosov#0974"

@evgeniikozhanov

evgeniikozhanov commented:
> Попробовал запросить judgement, получил вот такую ошибку:
> 
> ```
> createType(Call):: Call: failed decoding identity.setIdentity:: Struct: failed on args: {"info":"{\"additional\":\"Vec<(Data,Data)>\",\"display\":\"Data\",\"legal\":\"Data\",\"web\":\"Data\",\"riot\":\"Data\",\"email\":\"Data\",\"pgpFingerprint\":\"Option<[u8;20]>\",\"image\":\"Data\",\"twitter\":\"Data\"}"}:: Struct: failed on info: {"additional":"Vec<(Data,Data)>","display":"Data","legal":"Data","web":"Data","riot":"Data","email":"Data","pgpFingerprint":"Option<[u8;20]>","image":"Data","twitter":"Data"}:: Struct: failed on display: {"_enum":{"None":"Null","Raw":"Bytes","BlakeTwo256":"H256","Sha256":"H256","Keccak256":"H256","ShaThree256":"H256"}}:: Unable to decode Data, invalid indicator byte 116
> ```
> 
> Выбирал второго регистратора (Caldera), заполнял поля display name "test identity" и поле discord "fedosov#0974"
> 
> @evgeniikozhanov

I reformatted the request structure.

fedosov commented:
Не передалось значение из поля Discord. Как тестировал:
завел новый аккаунт
запросил идентити
заполнял поля display name "test identity" и поле discord "fedosov#0974"
у этого аккаунта не заполнено поле discord, только name записался

Проверять поля можно тут: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc2.3dpass.org#/accounts
Нажать на аккаунт и справа будет поле Identity. Там же и должно быть поле DISCORD, если было указано при создании заявки

evgeniikozhanov commented:
> Не передалось значение из поля Discord. Как тестировал: завел новый аккаунт запросил идентити заполнял поля display name "test identity" и поле discord "fedosov#0974" у этого аккаунта не заполнено поле discord, только name записался
> 
> Проверять поля можно тут: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc2.3dpass.org#/accounts Нажать на аккаунт и справа будет поле Identity. Там же и должно быть поле DISCORD, если было указано при создании заявки

should be fixed

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -108,7 +108,11 @@ export default function DialogIdentity({ isOpen, onClose, pair, hasIdentity }: I
       const candidateData = {} as PalletIdentityIdentityInfo;
       for (const [key, value] of Object.entries(dataState.candidateInfo)) {
-        candidateData[key] = {Raw: value};
+        if (key == 'discord') {
```
Похоже что у тебя форматирование слетело, вроде настраивали ранее

evgeniikozhanov replied:
+

fedosov commented:
Дискорд прописался, но другая проблема осталась.

![image](https://github.com/3Dpass/wallet/assets/556770/acc43e68-b757-45b1-a3a6-78ebb905f611)

Мне пишут что requestJudgement экстринсик не отправился в сеть. Он отправляется вообще?

UPD: спустя пару минут у меня появилось окно про эту отправку. Это совсем неинтуитивно. Надо давать пользователю знать что происходит. Я в курсе процесса и все равно я запутался. Должен быть хотя бы какой-то индикатор, что процесс идет и надо ждать, не уходить из кошелька.

evgeniikozhanov commented:
> Дискорд прописался, но другая проблема осталась.
> 
> ![image](https://user-images.githubusercontent.com/556770/237821938-acc43e68-b757-45b1-a3a6-78ebb905f611.png)
> 
> Мне пишут что requestJudgement экстринсик не отправился в сеть. Он отправляется вообще?
> 
> UPD: спустя пару минут у меня появилось окно про эту отправку. Это совсем неинтуитивно. Надо давать пользователю знать что происходит. Я в курсе процесса и все равно я запутался. Должен быть хотя бы какой-то индикатор, что процесс идет и надо ждать, не уходить из кошелька.

I added a logic that we show a "success" message only if the second request status is "ready". And I added a warning message for the first request.

fedosov commented on app/components/dialogs/DialogIdentity.tsx:
```diff
@@ -123,15 +123,25 @@ export default function DialogIdentity({ isOpen, onClose, pair, hasIdentity }: I
+        icon: "warning-sign",
+        intent: Intent.WARNING,
+        message: "Your request is in progress, please stay on for a while",
+        timeout: 20000,
```
Давай еще индикатор загрузки добавим, он очень наглядный. Смотри пример, как я это сделал для создания пула (`isCreatePoolLoading`)

evgeniikozhanov replied:
but there are no buttons here... should the dialog be kept opened in the loading case?

fedosov replied:
Да, ты прав, диалоговое окно должно оставаться открытым до окончания операции. В этом окне есть же кнопка? Вот ее можно сделать loading

evgeniikozhanov replied:
done

fedosov reviewed (commented):
Буду тестировать, спасибо!

fedosov commented:
Во время теста получил такую ошибку:

Uncaught (in promise) TypeError: Cannot read properties of undefined (reading '1')
    at Account.client.tsx:109:46

<img width="702" alt="image" src="https://github.com/3Dpass/wallet/assets/556770/ae2d5cf4-27d9-419b-8929-4411acac7025">

Данные заполнял такие же как и всегда (описал выше), регистратора выбирал Caldera

evgeniikozhanov commented:
> Во время теста получил такую ошибку (завис на окне и вечная загрузка):
> 
> ```
> Uncaught (in promise) TypeError: Cannot read properties of undefined (reading '1')
>     at Account.client.tsx:109:46
> ```
> 
> <img alt="image" width="702" src="https://user-images.githubusercontent.com/556770/238765025-ae2d5cf4-27d9-419b-8929-4411acac7025.png">
> Данные заполнял такие же как и всегда (описал выше), регистратора выбирал Caldera

Ah, yes, I faced this one, but then I forgot about it. I added a condition to prevent using undefined in this clause.
